### DISCUSSION TITLE
Testing concurrent requests

### DISCUSSION THREAD

**Discussion by nandi95** [laravel/framework#48623] [Ideas]
I have some code sensitive to a race condition, I added in some logic to lock with the cache, but I do not know how to test this automatically. Can we have some logic added to `src/Illuminate/Foundation/Testing/Concerns/MakesHttpRequests.php` so it allows "sending" multiple requests? I imagine the interface would be similar to `Http::pool` perhaps?

**Comment by AliBayat** [comment]
I'm facing the exact situation, would be nice if some wrapper can add that to native phpunit http methods

**Reply by AliBayat** [reply]
since my issue was arising as I was executing things in a dispatched job,
so far I have come up with the following workaround on the testing matter:

- setup an asynchronous job driver handler (rabbitMQ in my case) for the testing environment and set it up in .env.testing
- asserting the outcome of dispatching jobs without race conditions
- asserting that multiple requests to a certain part of the system get only one 200-status

**note**: this is not an ideal solution as I rather use a sync driver for my testing pipeline

**Comment by HasanMothaffar** [comment]
I'm in the same situation but using `lockForUpdate`. I thought of using `HTTP:pool` but on this page https://laravel.com/docs/10.x/http-tests#validation-assertions it says this:

"In general, each of your tests should only make one request to your application. Unexpected behavior may occur if multiple requests are executed within a single test method."

**Reply by nandi95** [reply]
I don't know what kind of unexpected behaviors there could be as side effects in the logic are expected.
`Http::pool` would make an actual request, whereas when you run `$this->postJson(...` it will emulate the application by creating a request and running the request through the app as it normally would.

**Comment by mityukov** [comment]
Considering using Concurrency facade instead of Http::pool() inside a test. This way I could call specific service, avoiding "full-blown" http requests. Would it fly?
